tools/unirestore/README.md

```diff
@@ -0,0 +1,99 @@
+# Universal Restore Utility
+
+Restore a backup of a set of SSTable files taken from any given Scylla/Cassandra cluster to any size/topology cluster. 
+
+
+Overview of what it does and how it does it:
+
+As a user, I have a set of sstable files I have backed up from my old cluster. The set looks like this[1]:
+
+`root-dir/host/keyspace/table/sstable_files`
+
+Also, the user provides me with:
+
+- the original schema dump in a CQL file (can be sanitized to include only keyspaces/tables)
+- ssh credentials for a sudo user on all nodes (key and password access are supported)
+- One of the node IPs in the cluster for cluster metadata gathering
+- If needed, API URL and CQLSH credentials
+
+I have a new cluster into which I'd like to load the data. The node count and topology is irrelevant (though if the new cluster 
+is multi-DC based, it would make sense to limit the destination node list to a local DC and adjust the replicas after restore, 
+in order to avoid cross DC traffic)
+
+I assume the schema has already been loaded into the new cluster.
+
+## The workflow:
+
+1. Parse the original schema dump, to have a list of valid keyspaces and tables.
+2. Extract the cluster node IP list. 
+3. Ensure schema versions are up to date, and in sync across the cluster.
+4. Extract the token ring data for each valid keyspace [temporarily unavailable]
+
+5. Find the correct destination paths for each ks/table (use system tables to avoid looking at dropped tables)
+  5.1 Ensure the destination dirs are writeable
+6. Build restore plan
+  6.1 organize source files as bundles of 9 sstable files with source paths
+  6.2 [temporarily unavailable] add the first/last tokens set for each bundle
+  6.3 [temporarily unavailable] add the list of hosts for each keyspace where the file token range overlaps the node/ks token range
+  6.4 add destination path (ks/table/upload) to each bundle
+7. Iterate over the bundled file sets, go one source host per iteration, copying the largest bundles first.
+  7.1 Measure the bundle (9 sstable files) size
+  7.2 Check `df` on the destination nodes
+    7.2.1 Confirm There's room for the incoming bundle
+    7.2.2 Make sure the preset percentage of free space is not exceeded
+    7.2.3 If disk space is low or below the threshold
+      7.2.3.1 Send the API call to stop compactions
+      7.2.3.2 Set the correct file permissions on the data in the upload dirs
+      7.2.3.3 Call nodetool refresh[2], and wait for it to finish[3]
```

Comment:
I don't understand how stopping compactions solves the issue of disk space.

I guess there is either some techincal part I am missing as to why this helps, or assumed to help.

Please note that after nodetool refresh - there can be many compactions running trying to reduce data duplicates - stopping compactions doesn't help

Reply:
> In some cases we will need to run nodetool compact as well to reduce space amplification due to duplicates (we are 
> copying all the files from the source cluster RF copies to a single node). If you have it somewhere I have missed it.

Not doing it now, because compactions interfere with uploads (the lock problem Glauber was working on). I actually issue a stop to currently ongoing compactions to mitigate the issue (I know it doesn't get rid of it completely). Stopping compactions until the restore is done was actually suggested as a workaround for the resurrection issue @avikivity brough up. 

> If the original cluster has been doing nodetool compact with STCS - you can expect huge files and in this case I can see 
> how this procedure will not work if there is a major single table with the majority of the disk space usage (compared to 
> others).

The script checks `df` on the node/mountpoint for free space before uploading each "bundle" (bundle == set of sstables with the same index). If the node doesn't have enough space *2 to contain the whole bundle and if the node will go over the allowed free space quota (in percent) after the upload, the process stops with an error. 
If you can suggest an heuristic to also trigger a local compaction in there, I'll add it to the script as well

> Consider the extreme case of 9 nodes with single shard with RF=3 using STCS (src cluster) - copying to another cluster 
> with 9 nodes with single shard and RF=3 (dest cluster). SRC does nodetool compact everyday, and nodes are at 50% 
> disk space. So you have on every node of src cluster a single sstable to copy taking 50% of the disk size.
>
> The procedure will not work in my view for this case.

Absolutely. @nyh mentioned a way to split sstables before, so there might be a workaround, but I think that would be out of the current scope.

> One item that may help here is if we add to the nodetool refresh the ability to cleanup the data (while 
> reshaping / resharding) - but we do not have it today.

The script runs refresh and then cleanup (once the upload dir is empty). I guess this is as good as it can be with the current Scylla capabilities

Reply:
> I don't understand how stopping compactions solves the issue of disk space.
> I guess there is either some techincal part I am missing as to why this helps, or assumed to help.

It doesn't help with disk space, this is an attempt (not guaranteed of course) to mitigate the compaction lock issue and avoid resurrections

> Please note that after nodetool refresh - there can be many compactions running trying to reduce data duplicates - stopping compactions doesn't help

Is there anything else I can add here to improve the situation?